File: Server/import-products.py
"""
Product Import Script
Imports engagement ring products from Excel file into the database
"""

import openpyxl
import sys
import os
import psycopg2
from psycopg2.extras import RealDictCursor
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Database connection
DB_CONFIG = {
    'host': os.getenv('PG_HOST', 'localhost'),
    'port': os.getenv('PG_PORT', 5432),
    'database': os.getenv('PG_DATABASE'),
    'user': os.getenv('PG_USERNAME'),
    'password': os.getenv('PG_PASSWORD')
}

def get_db_connection():
    """Create database connection"""
    return psycopg2.connect(**DB_CONFIG)

def get_category_id(cursor, category_name='Rings'):
    """Get category ID by name"""
    cursor.execute("SELECT id FROM categories WHERE name = %s", (category_name,))
    result = cursor.fetchone()
    return result['id'] if result else None

def get_jewelry_sub_type_id(cursor, sub_type_name='Engagement Rings'):
    """Get jewelry sub type ID"""
    cursor.execute("SELECT id FROM jewelry_sub_types WHERE name = %s", (sub_type_name,))
    result = cursor.fetchone()
    return result['id'] if result else None

def get_ring_style_mapping(cursor):
    """Get all ring styles and create name->id mapping"""
    cursor.execute("SELECT id, name FROM ring_types")
    return {row['name']: row['id'] for row in cursor.fetchall()}

def get_stone_shape_mapping(cursor):
    """Get all stone shapes and create name->id mapping"""
    cursor.execute("SELECT id, name FROM stone_shapes")
    return {row['name']: row['id'] for row in cursor.fetchall()}

def get_stone_type_mapping(cursor):
    """Get all stone types and create name->id mapping"""
    cursor.execute("SELECT id, name FROM stone_types")
    return {row['name']: row['id'] for row in cursor.fetchall()}

def product_exists(cursor, sku):
    """Check if product with SKU already exists"""
    cursor.execute("SELECT id FROM products WHERE sku = %s", (sku,))
    return cursor.fetchone() is not None

def generate_slug(name):
    """Generate URL-friendly slug from product name"""
    return name.lower().replace(' ', '-').replace("'", '')

def import_products():
    """Main import function"""
    excel_file = r'C:\Users\minuk\Videos\Diamond Ring Set_Aravinth.xlsx'

    print('='*80)
    print('PRODUCT IMPORT SCRIPT')
    print('='*80)
    print()

    # Load Excel file
    print('Loading Excel file...')
    wb = openpyxl.load_workbook(excel_file, data_only=True)

    # Find Engagement Ring sheet
    engagement_sheet = None
    for sheet_name in wb.sheetnames:
        if 'engagement' in sheet_name.lower():
            engagement_sheet = wb[sheet_name]
            print(f'Found sheet: {sheet_name}')
            break

    if not engagement_sheet:
        print('ERROR: Engagement Ring sheet not found!')
        return

    print(f'Sheet has {engagement_sheet.max_row - 1} data rows')
    print()

    # Connect to database
    print('Connecting to database...')
    conn = get_db_connection()
    cursor = conn.cursor(cursor_factory=RealDictCursor)

    try:
        # Get mappings
        print('Loading database mappings...')
        category_id = get_category_id(cursor)
        jewelry_sub_type_id = get_jewelry_sub_type_id(cursor)
        ring_style_map = get_ring_style_mapping(cursor)
        stone_shape_map = get_stone_shape_mapping(cursor)
        stone_type_map = get_stone_type_mapping(cursor)

        if not category_id:
            print('ERROR: Rings category not found!')
            return

        if not jewelry_sub_type_id:
            print('ERROR: Engagement Rings sub-type not found!')
            return

        print(f'  - Category ID: {category_id}')
        print(f'  - Jewelry Sub Type ID: {jewelry_sub_type_id}')
        print(f'  - Ring Styles: {len(ring_style_map)} loaded')
        print(f'  - Stone Shapes: {len(stone_shape_map)} loaded')
        print(f'  - Stone Types: {len(stone_type_map)} loaded')
        print()

        # Process each row
        imported_count = 0
        skipped_count = 0
        error_count = 0

        for row_idx in range(2, engagement_sheet.max_row + 1):
            row = engagement_sheet[row_idx]

            # Read row data
            sku = str(row[0].value).strip() if row[0].value else None
            name = str(row[2].value).strip() if row[2].value else None
            description = str(row[3].value).strip() if row[3].value else None
            ring_style_1 = str(row[4].value).strip() if row[4].value else None
            ring_style_2 = str(row[5].value).strip() if row[5].value else None
            ring_style_3 = str(row[6].value).strip() if row[6].value else None
            ring_style_4 = str(row[7].value).strip() if row[7].value else None
            ring_style_5 = str(row[8].value).strip() if row[8].value else None
            stone_shape = str(row[9].value).strip() if row[9].value else None
            stone_type = str(row[10].value).strip() if row[10].value else None

            # Skip rows without SKU or name
            if not sku or not name or name == '#N/A':
                skipped_count += 1
                continue

            # Check if product already exists
            if product_exists(cursor, sku):
                print(f'  [{row_idx-1}/{engagement_sheet.max_row-1}] SKIPPED: {sku} - {name} (already exists)')
                skipped_count += 1
                continue

            try:
                # Map ring styles to IDs
                ring_style_ids = []
                for style_name in [ring_style_1, ring_style_2, ring_style_3, ring_style_4, ring_style_5]:
                    if style_name and style_name in ring_style_map:
                        ring_style_ids.append(ring_style_map[style_name])
                    else:
                        ring_style_ids.append(None)

                # Map stone shape to ID
                stone_shape_id = stone_shape_map.get(stone_shape) if stone_shape else None

                # Map stone type to ID
                stone_type_id = stone_type_map.get(stone_type) if stone_type else None

                # Generate slug
                slug = generate_slug(name)

                # Set default price (you can adjust this)
                base_price = 1000.00

                # Insert product
                cursor.execute("""
                    INSERT INTO products (
                        id, name, slug, description, short_description, sku,
                        base_price, currency, category_id, jewelry_sub_type_id,
                        stone_type_id, ring_style_1_id, ring_style_2_id, ring_style_3_id,
                        ring_style_4_id, ring_style_5_id,
                        is_active, is_featured, in_stock, stock_quantity,
                        meta_title, meta_description,
                        created_at, updated_at
                    ) VALUES (
                        gen_random_uuid(), %s, %s, %s, %s, %s,
                        %s, %s, %s, %s,
                        %s, %s, %s, %s, %s, %s,
                        %s, %s, %s, %s,
                        %s, %s,
                        NOW(), NOW()
                    ) RETURNING id
                """, (
                    name, slug, description, description[:200] if description else '', sku,
                    base_price, 'GBP', category_id, jewelry_sub_type_id,
                    stone_type_id, ring_style_ids[0], ring_style_ids[1], ring_style_ids[2],
                    ring_style_ids[3], ring_style_ids[4],
                    True, False, True, 1,
                    name, description[:160] if description else ''
                ))

                product_id = cursor.fetchone()['id']

                # Add stone shape relationship if available
                if stone_shape_id:
                    cursor.execute("""
                        INSERT INTO product_stone_shapes (id, product_id, stone_shape_id, created_at, updated_at)
                        VALUES (gen_random_uuid(), %s, %s, NOW(), NOW())
                    """, (product_id, stone_shape_id))

                print(f'  [{row_idx-1}/{engagement_sheet.max_row-1}] IMPORTED: {sku} - {name}')
                imported_count += 1

            except Exception as e:
                print(f'  [{row_idx-1}/{engagement_sheet.max_row-1}] ERROR: {sku} - {name}: {str(e)}')
                error_count += 1
                continue

        # Commit all changes
        conn.commit()

        # Print summary
        print()
        print('='*80)
        print('IMPORT SUMMARY')
        print('='*80)
        print(f'Total rows processed: {engagement_sheet.max_row - 1}')
        print(f'Successfully imported: {imported_count}')
        print(f'Skipped (already exists or invalid): {skipped_count}')
        print(f'Errors: {error_count}')
        print()
        print('Import completed successfully!')

    except Exception as e:
        conn.rollback()
        print(f'\nFATAL ERROR: {str(e)}')
        import traceback
        traceback.print_exc()
    finally:
        cursor.close()
        conn.close()

if __name__ == '__main__':
    import_products()

File: Server/import-products.js
const { Sequelize } = require('sequelize');
const XLSX = require('xlsx');
const path = require('path');
const fs = require('fs');

const sequelize = new Sequelize({
  host: '31.97.116.89',
  port: 5432,
  database: 'mcculloch_db',
  username: 'mcculloch_admin',
  password: '#mcculloch_admin#20026',
  dialect: 'postgres',
  logging: false
});

// Configuration
const BATCH_SIZE = 10; // Process 10 products at a time
const DEFAULT_PRICE = 1000; // Default price if not specified
const DEFAULT_CURRENCY = 'USD';

// Helper function to create slug from name
function createSlug(text) {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)/g, '');
}

// Helper function to get lookup ID
async function getLookupId(tableName, valueName) {
  if (!valueName || String(valueName).trim() === '' || String(valueName) === '#N/A') {
    return null;
  }

  const cleanValue = String(valueName).trim();

  try {
    const [results] = await sequelize.query(
      `SELECT id FROM "${tableName}" WHERE name = :name LIMIT 1`,
      { replacements: { name: cleanValue } }
    );

    if (results.length > 0) {
      return results[0].id;
    }

    console.warn(`   ⚠️  Warning: "${cleanValue}" not found in ${tableName}`);
    return null;
  } catch (error) {
    console.error(`   ❌ Error looking up "${cleanValue}" in ${tableName}:`, error.message);
    return null;
  }
}

// Validate and prepare product data
async function prepareProductData(row, index) {
  // Handle missing or invalid names
  const code = row['code '] ? String(row['code ']).trim() : null;
  let name = row['Name'] ? String(row['Name']).trim() : null;
  let description = row['Description'] ? String(row['Description']).trim() : null;

  // Generate test name/description if missing
  if (!name || name === '' || name === '#N/A') {
    name = `Test Name ${index + 1}`;
    console.log(`   ⚠️  Row ${index + 1}: Generated name "${name}"`);
  }

  if (!description || description === '' || description === '#N/A') {
    description = `Test Description ${index + 1}`;
    console.log(`   ⚠️  Row ${index + 1}: Generated description`);
  }

  // Create slug
  const slug = createSlug(name) + (code ? `-${code.toLowerCase()}` : `-${index + 1}`);

  // Get lookup IDs
  const ringStyle1Id = await getLookupId('ring_types', row['Ring Style-1']);
  const ringStyle2Id = await getLookupId('ring_types', row['Ring Style-2']);
  const ringStyle3Id = await getLookupId('ring_types', row['Ring Style-3']);
  const ringStyle4Id = await getLookupId('ring_types', row['Ring Style-4']);
  const stoneShapeId = await getLookupId('stone_shapes', row['Stone Shape']);
  const stoneTypeId = await getLookupId('stone_types', row['Stone Type']);

  return {
    sku: code || `SKU-${index + 1}`,
    name: name,
    slug: slug,
    description: description,
    base_price: DEFAULT_PRICE,
    currency: DEFAULT_CURRENCY,
    is_active: true,
    is_featured: false,
    in_stock: true,
    ring_style_1_id: ringStyle1Id,
    ring_style_2_id: ringStyle2Id,
    ring_style_3_id: ringStyle3Id,
    ring_style_4_id: ringStyle4Id,
    ring_style_5_id: null, // Not in Excel
    stone_shape_id: stoneShapeId,
    stone_type_id: stoneTypeId
  };
}

// Import products in batches
async function importProducts() {
  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  try {
    await sequelize.authenticate();
    console.log('✓ Connected to database\n');

    // Read Excel file
    const excelPath = path.join(__dirname, '..', 'Product-list.xlsx');
    const workbook = XLSX.readFile(excelPath);
    const worksheet = workbook.Sheets[workbook.SheetNames[0]];
    const data = XLSX.utils.sheet_to_json(worksheet);

    console.log('═══════════════════════════════════════════');
    console.log('    PRODUCT IMPORT - BATCH PROCESSING');
    console.log('═══════════════════════════════════════════\n');
    console.log(`📦 Total products to import: ${data.length}`);
    console.log(`📊 Batch size: ${BATCH_SIZE} products per batch`);
    console.log(`📈 Total batches: ${Math.ceil(data.length / BATCH_SIZE)}\n`);

    // Process in batches
    for (let i = 0; i < data.length; i += BATCH_SIZE) {
      const batch = data.slice(i, i + BATCH_SIZE);
      const batchNumber = Math.floor(i / BATCH_SIZE) + 1;
      const totalBatches = Math.ceil(data.length / BATCH_SIZE);

      console.log(`\n╔═══════════════════════════════════════════╗`);
      console.log(`║   BATCH ${batchNumber}/${totalBatches} (Products ${i + 1}-${Math.min(i + BATCH_SIZE, data.length)})      `);
      console.log(`╚═══════════════════════════════════════════╝\n`);

      for (let j = 0; j < batch.length; j++) {
        const row = batch[j];
        const rowNumber = i + j + 1;

        try {
          console.log(`📦 Processing Product ${rowNumber}/${data.length}...`);

          // Prepare data with validation
          const productData = await prepareProductData(row, i + j);

          // Insert into database
          await sequelize.query(`
            INSERT INTO products (
              sku, name, slug, description, base_price, currency,
              is_active, is_featured, in_stock,
              ring_style_1_id, ring_style_2_id, ring_style_3_id, ring_style_4_id, ring_style_5_id,
              stone_shape_id, stone_type_id,
              created_at, updated_at
            ) VALUES (
              :sku, :name, :slug, :description, :base_price, :currency,
              :is_active, :is_featured, :in_stock,
              :ring_style_1_id, :ring_style_2_id, :ring_style_3_id, :ring_style_4_id, :ring_style_5_id,
              :stone_shape_id, :stone_type_id,
              CURRENT_TIMESTAMP, CURRENT_TIMESTAMP
            )
          `, {
            replacements: productData
          });

          console.log(`   ✅ SUCCESS: "${productData.name}" (SKU: ${productData.sku})`);
          successCount++;

        } catch (error) {
          console.error(`   ❌ ERROR: Product ${rowNumber} failed - ${error.message}`);
          errors.push({
            row: rowNumber,
            data: row,
            error: error.message
          });
          errorCount++;
        }
      }

      // Small delay between batches
      if (i + BATCH_SIZE < data.length) {
        console.log(`\n   ⏳ Batch complete. Pausing for 1 second...\n`);
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
    }

    // Final summary
    console.log('\n\n═══════════════════════════════════════════');
    console.log('    IMPORT COMPLETE - SUMMARY');
    console.log('═══════════════════════════════════════════\n');
    console.log(`✅ Successfully imported: ${successCount} products`);
    console.log(`❌ Failed: ${errorCount} products`);
    console.log(`📊 Success rate: ${((successCount / data.length) * 100).toFixed(2)}%\n`);

    if (errors.length > 0) {
      console.log('❌ ERRORS:\n');
      errors.forEach(err => {
        console.log(`   Row ${err.row}: ${err.error}`);
      });

      // Save errors to file
      fs.writeFileSync(
        path.join(__dirname, 'import-errors.json'),
        JSON.stringify(errors, null, 2)
      );
      console.log('\n💾 Errors saved to: import-errors.json\n');
    }

    // Verify database count
    const [count] = await sequelize.query('SELECT COUNT(*) as total FROM products');
    console.log(`📦 Total products in database: ${count[0].total}\n`);

    console.log('═══════════════════════════════════════════\n');

    await sequelize.close();

  } catch (error) {
    console.error('\n❌ FATAL ERROR:', error.message);
    console.error(error);
    process.exit(1);
  }
}

// Run import
importProducts();
